### ISSUE TITLE
nvim_command edit API blocks on swap recovery dialog

### ISSUE THREAD

**Issue by battlmonstr** [neovim/neovim#7142]
- `nvim --version`: 0.2.0
- Vim (version: ) behaves differently? N/A
- Operating system/version: macOS Sierra 10.12.1
- Terminal name/version: a remote UI talking to a --headless nvim through the Nvim API
- `$TERM`: N/A

### Steps to reproduce using `nvim -u NORC`

Run:
    
    NVIM_LISTEN_ADDRESS=... nvim -u NORC --headless 

Connect to that address, and issue an API call:

    nvim_command("edit README.md")

Kill the nvim process and restart it.
Reconnect and issue the same command.

Observe that nvim prints out a swap file recovery dialog with a prompt to nvim's stdout:

```
Found a swap file by the name ...
...
[O]pen Read-Only, (E)dit anyway, (R)ecover, (D)elete it, (Q)uit, (A)bort: 
```

**Note**: Changing `nvim_command` to `nvim_command_output` exposes the same behaviour.

### Actual behaviour

The API call blocks and never returns, neither success nor failure.
Moreover, no API calls work from this point.
For example, calling `nvim_command("recover")` or `nvim_command("help recovery")` blocks and doesn't return any result either, and has no effect on nvim.

### Expected behaviour

`nvim_command` must not block. It should return immediately an error reply explaining that a recovery decision is required. Until the decision is taken, other commands like `nvim_command("help recovery")` should be processed normally.

**Comment by battlmonstr** [comment]
The problem seems to not be special to recovery or "edit". It happens also with some other commands that provoke dialogs & prompts.

For example, `nvim_command("version")` provokes a prompt:

> Press ENTER or type command to continue

and never returns. Further `nvim_command` calls are blocked In this state.

This might be mitigated by fixing https://github.com/neovim/neovim/issues/6201

**Comment by battlmonstr** [comment]
Possibly a duplicate of:
https://github.com/neovim/neovim/issues/3901

**Comment by justinmk** [comment]
Only "async" API functions are non-blocking. Search `:help api` for "async" to see which functions those are (there's only ~3).

All other API functions block while nvim is waiting for input. To avoid being blocked in a client, you can use `nvim_get_mode()` which returns a `blocking` flag. 

You can send input to nvim (just as a tty/GUI user could send keyboard input) via the async `nvim_input` function. This can be used to unblock nvim.

Future improvements are planned:

1. Add more async API functions
2. Offer a mode where functions return with an error if they would block
3. Reduce the cases where nvim blocks for a long time.

**Comment by battlmonstr** [comment]
Thank you for the detailed reply.
Having such a mode, which returns an error if vim is blocked waiting for input, would be very nice.
It would be nice if the initial call (that provoked that) also returned a "blocked on a dialog" error.

I guess that externalized dialogs would solve the last bit of the problem where I know it's blocked, but I don't really know what to send to nvim_input to unblock the engine.

**Comment by battlmonstr** [comment]
I don't quite understand an idea about a timeout. If the engine is waiting for input this should be known immediately in the engine somehow without a delay (in most cases). Someting like "error: blocked on a dialog input" makes sense here. A timeout would probably complicate things.

**Comment by justinmk** [comment]
> I don't quite understand an idea about a timeout. If the engine is waiting for input this should be known immediately in the engine somehow without a delay (in most cases).

Nvim knows it's blocked, but maybe it's only blocked for a 10 ms. In that case rather than return an error, it would wait 10ms then proceed as normal.
